src/lib/child_processes/child_processes.ml

```diff
@@ -0,0 +1,444 @@
+          let open Deferred.Let_syntax in
+          let%bind process =
+            start_custom ~logger ~name ~checkout_relative_path ~conf_dir
+              ~args:["exit"]
+              ~stdout:(`Log Logger.Level.Debug, `Pipe)
+              ~stderr:(`Log Logger.Level.Error, `No_pipe)
+              ~termination:`Raise_on_failure
+            |> Deferred.map ~f:Or_error.ok_exn
+          in
+          let%bind () =
+            Strict_pipe.Reader.iter (stdout_lines process) ~f:(fun line ->
+                [%test_eq: string] "hello" line ;
+                Deferred.unit )
+          in
+          (* Pipe will be closed before the ivar is filled, so we need to wait a
+             bit. *)
+          let%bind () = after @@ Time.Span.of_sec 0.1 in
+          [%test_eq: Unix.Exit_or_signal.t option] (Some (Ok ()))
+            (termination_status process) ;
+          Deferred.unit )
+
+    let%test_unit "killing works" =
+      async_with_temp_dir (fun conf_dir ->
+          let open Deferred.Let_syntax in
+          let%bind process =
+            start_custom ~logger ~name ~checkout_relative_path ~conf_dir
+              ~args:["loop"]
+              ~stdout:(`Don't_log, `Pipe)
+              ~stderr:(`Don't_log, `No_pipe)
+              ~termination:`Always_raise
+            |> Deferred.map ~f:Or_error.ok_exn
+          in
+          let lock_exists () =
+            Deferred.map
+              (Sys.file_exists (conf_dir ^/ name ^ ".lock"))
+              ~f:(function `Yes -> true | _ -> false)
+          in
+          let assert_lock_exists () =
+            Deferred.map (lock_exists ()) ~f:(fun exists -> assert exists)
+          in
+          let assert_lock_does_not_exist () =
+            Deferred.map (lock_exists ()) ~f:(fun exists -> assert (not exists))
+          in
+          let%bind () = assert_lock_exists () in
+          let output = ref [] in
+          let rec go () =
+            match%bind Strict_pipe.Reader.read (stdout_lines process) with
+            | `Eof ->
+                failwith "pipe closed when process should've run forever"
+            | `Ok line ->
+                output := line :: !output ;
+                if List.length !output = 10 then Deferred.unit else go ()
+          in
+          let%bind () = go () in
+          [%test_eq: string list] !output (List.init 10 ~f:(fun _ -> "hello")) ;
+          let%bind () = after @@ Time.Span.of_sec 0.1 in
+          assert (Option.is_none @@ termination_status process) ;
+          let%bind kill_res = kill process in
+          let%bind () = assert_lock_does_not_exist () in
+          let exit_or_signal = Or_error.ok_exn kill_res in
+          [%test_eq: Unix.Exit_or_signal.t] exit_or_signal
+            (Error (`Signal Signal.term)) ;
+          assert (Option.is_some @@ termination_status process) ;
+          Deferred.unit )
+
+    let () =
+      Async.Scheduler.set_record_backtraces true ;
+      Core.Backtrace.elide := false
```

Comment:
remove these or put them at the top of the module imo?

Reply:
yeah, done